### PR TITLE
Refactor fetching function

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

The previous implementations had some code smells. This PR simplifies it.

Part of #11515

### Does your PR introduce any breaking changes? If yes, please list them.

None user-facing

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @justusschock @awaelchli @akihironitta @rohitgr7 @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/utilities/fetching.py [hunk 1]
```diff
@@ -228,44 +228,33 @@ def wait(self) -> None:
         """Hook to override to indicate the `DataFetcher` to wait for an event."""
 
     def prefetching(self) -> None:
+        iterator = self.dataloader_iter
         for _ in range(self.prefetch_batches):
             try:
-                self._fetch_next_batch()
+                self._fetch_next_batch(iterator)
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:234]:
```suggestion
                self._fetch_next_batch(self.dataloader_iter)
```
No need for a separate variable :)

File: pytorch_lightning/utilities/fetching.py [hunk 2]
```diff
@@ -228,44 +228,33 @@ def wait(self) -> None:
         """Hook to override to indicate the `DataFetcher` to wait for an event."""
 
     def prefetching(self) -> None:
+        iterator = self.dataloader_iter
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:231]:
```suggestion
```

File: pytorch_lightning/utilities/fetching.py [hunk 3]
```diff
@@ -228,44 +228,33 @@ def wait(self) -> None:
...
-        if self.done:
-            while self.batches:
-                return self._get_queued_batch()
-            raise StopIteration
+    def fetching_function(self) -> Tuple[Any, bool]:
+        if self.batches:
+            batch = self.batches.pop(0)
         else:
+            # empty iterator, no prefetching done
+            raise StopIteration
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:243]: This means, you won't get a batch here at all right? If no pre-fetching was done, we should still wait for a batch to be normally loaded, no? How can this work if you early-exit the function here? The `__iter__` imoplementation does not catch this and therefore it would result in end of iterating at all or am I missing something?

File: pytorch_lightning/utilities/fetching.py [hunk 4]
```diff
@@ -228,44 +228,33 @@ def wait(self) -> None:
...
+                self._fetch_next_batch(self.dataloader_iter)
             except StopIteration:
-                self.batches.insert(0, yield_batch)
                 self.done = True
-                return self._get_queued_batch()
-
-            except IndexError:
-                raise StopIteration
+        self.wait()
+        return self.move_to_device(batch), len(self.batches) == 0
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:251]: How do you get this batch in case of no pre-fetching?

Comment (carmocca) [reply pytorch_lightning/utilities/fetching.py:251]: I'll follow-up in your other comment which I think asks the same thing

File: pytorch_lightning/utilities/fetching.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/utilities/fetching.py:243]: pre-fetching=0 is not supported. See the original implementation before I started refactoring: https://github.com/PyTorchLightning/pytorch-lightning/blob/0066ff012962ffa920e9bbd750f4edbea05d4505/pytorch_lightning/utilities/fetching.py#L65-L71

This should be explicitly checked for, but needs further refactoring of the inheritance structure.

The only class that uses `prefetch_batches=0` is `DataLoaderIterDataFetcher` but it overrides `fetching_function`.

**Review by tchaton — APPROVED** [review]
Nice !

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/utilities/fetching.py [hunk 5]
```diff
@@ -228,44 +228,33 @@ def wait(self) -> None:
...
-                return self._get_queued_batch()
-            raise StopIteration
+    def fetching_function(self) -> Tuple[Any, bool]:
+        if self.batches:
+            batch = self.batches.pop(0)
         else:
+            # empty iterator, no prefetching done
+            raise StopIteration
+        if not self.done:
+            assert self.dataloader_iter is not None
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/fetching.py:245]: Maybe a `RuntimeError` or `AssertionError` would be better here, with a message saying to open an issue on github or somethingn (since presumably we don't expect this to ever fail)

**Review by ethanwharris — APPROVED** [review]
LGTM, small comment

File: pytorch_lightning/utilities/fetching.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/utilities/fetching.py:245]: `assert ...` will raise an `AssertionError`, no?

Note that I'm doing this just because of typing (#11515), we are forced to add many of these assertions throughout the codebase because optional set via callbacks are un-checkable with mypy.

Comment (ethanwharris) [reply pytorch_lightning/utilities/fetching.py:245]: Assert statements can be optimized away (not guaranteed to be run) so they're not generally recommend if there's any real chance they could fail. Since this is just to cheer up mypy then I think it's fine :smiley:
